### PR TITLE
Add dataset related function for multi worker/process distribution.

### PR DESCRIPTION
**Pull request by qlzh727**
1. Update the distribution class to aware of single/multi worker setting.
2. Add a new method in Distribution to shard the dataset. 
3. Added unit test for newly added methods.

TODO: 
1. Will have to add test coverage in the multi-worker setting internally.
2. Make sure the sharding are correct both data parallel (simple case), and model+data parallel (harder case).

This will address part of the #18561

### CODE REVIEW

File: keras/distribution/distribution_lib.py [hunk 1]
```diff
@@ -325,6 +325,18 @@ def scope(self):
     def device_mesh(self):
         return self._device_mesh
 
+    def distribute_dataset(self, tf_dataset):
```
Comment (fchollet) [inline keras/distribution/distribution_lib.py:328]: We need to make sure that we future-proof the API to deal with future evolution beyond tf.data (e.g. Grain). Here better to use the argument name `dataset` so we can extend to new types in the future.

File: keras/distribution/distribution_lib.py [hunk 2]
```diff
@@ -406,6 +421,14 @@ def get_tensor_layout(self, path):
         # For data parallel training, the intermediate state is not changed.
         return None
 
+    def distribute_dataset(self, tf_dataset):
+        if self._is_multi_process:
+            # TODO(scottzhu): Might want to add some prefetch for better perf.
```
Comment (fchollet) [inline keras/distribution/distribution_lib.py:426]: Maybe just call `prefetch` with autotune?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras/distribution/distribution_lib.py (hunk 1 above)
Comment (qlzh727) [reply keras/distribution/distribution_lib.py]: Done.

File: keras/distribution/distribution_lib.py (hunk 2 above)
Comment (qlzh727) [reply keras/distribution/distribution_lib.py]: Done.

**Review by fchollet — APPROVED** [review]
LGTM

**Comment by fchollet** [comment]
We may have to disabled `merging_test` for JAX too as well now.

**Pull request merged**